### PR TITLE
✨ Add proxy fields in CABPK

### LINKED ISSUE #3751: Http proxy support for egress traffic
**User Story**

As a operator I would like to proxy cluster egress traffic for security compliance
As a operator I would like to store proxy credentials in secret(s) and consume them securely
As a user I would like to update cluster proxy settings via cluster API
As a user I would like to view proxy configurations being applied for my cluster

**Detailed Description**

In some corporation network, security policies or firewall rules will be applied to block egress traffic to prevent data exfiltration, which will block the traffic for pulling images from external image registry or sending telemetry data. A proxy is required in this scenario to ensure the egress traffic can be managed in a secure way.

**Anything else you would like to add:**

With current **v1alpha3** implementation, we could potentially inject proxy configs to container runtime via the [**Files**](https://github.com/kubernetes-sigs/cluster-api/blob/27a6031a1d6aa1a310c905b7d1afa331ab1cfd2d/bootstrap/kubeadm/api/v1alpha2/kubeadmbootstrapconfig_types.go#L47) hook, however it has following problems:
1. It's making assumption that kubeadm/cloud-init is being provided.
2. You need to expose the credentials in the **kubeadmconfig** object.
3. It's not supporting reconfiguration. The [**Files**](https://github.com/kubernetes-sigs/cluster-api/blob/27a6031a1d6aa1a310c905b7d1afa331ab1cfd2d/bootstrap/kubeadm/api/v1alpha2/kubeadmbootstrapconfig_types.go#L47) or [**PreKubeadmCommands**](https://github.com/kubernetes-sigs/cluster-api/blob/27a6031a1d6aa1a310c905b7d1afa331ab1cfd2d/bootstrap/kubeadm/api/v1alpha2/kubeadmbootstrapconfig_types.go#L50) hook are making use of cloud-init and it will only take effective on bootstrap. You need to perform a machine deployment rolling upgrade to ensure the change is propagated to each VM within the deployment.

It'll be helpful to add configurations in the cluster API spec to natively support http proxy to fix the issues we have above.

/kind feature

### PR DESCRIPTION
**Pull request by hanlins**
**What this PR does / why we need it**:
Add proxy fields in CABPK

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #3751

### CODE REVIEW

File: bootstrap/kubeadm/api/v1alpha4/kubeadmbootstrapconfig_types.go [hunk 1]
```diff
@@ -333,5 +336,19 @@ type Filesystem struct {
...
 }
 
+// Proxy defines the http proxy settings used by container runtime during bootstrap
+type Proxy struct {
+	// HttpProxy specifies the proxy url for http traffic
+	HttpProxy string `json:"httpProxy,omitempty"`
+	// HttpsProxy specifies the proxy url for https traffic
+	HttpsProxy string `json:"httpsProxy,omitempty"`
+	// +optional
+	NoProxy []string `json:"noProxy,omitempty"`
```
Comment (fabriziopandini) [inline bootstrap/kubeadm/api/v1alpha4/kubeadmbootstrapconfig_types.go:346]: I think we should add comments to all the API fields

File: bootstrap/kubeadm/api/v1alpha4/kubeadmbootstrapconfig_types.go [hunk 2]
```diff
@@ -333,5 +336,19 @@ type Filesystem struct {
...
+	// +optional
+	HttpProxyAuth *string `json:"httpProxyAuth,omitempty"`
+	// +optional
+	HttpsProxyAuth *string `json:"httpsProxyAuth,omitempty"`
```
Comment (fabriziopandini) [inline bootstrap/kubeadm/api/v1alpha4/kubeadmbootstrapconfig_types.go:350]: Is this meant to store credentials?
If yes, what about transforming this into a link to a secret, so credentials don't get spread into all the kubeadm config objects (assuming they are the same) and it is possible to define RBAC around their access/management?

File: bootstrap/kubeadm/controllers/kubeadmconfig_controller.go [hunk 3]
```diff
@@ -563,6 +583,79 @@ func (r *KubeadmConfigReconciler) joinControlplane(ctx context.Context, scope *S
...
+	parsedUrl, err := url.Parse(proxyUrl)
+	if err != nil {
+		return "", err
+	}
+	parsedUrl.User = url.UserPassword(username, password)
+	return parsedUrl.String(), nil
+}
+
+// resolveProxyFiles resolves the required proxy unit files to cloudinit.Files
+func (r *KubeadmConfigReconciler) resolveProxyFiles(ctx context.Context, cfg *bootstrapv1.KubeadmConfig) ([]bootstrapv1.File, error) {
```
Comment (fabriziopandini) [inline bootstrap/kubeadm/controllers/kubeadmconfig_controller.go:602]: should this be 
```suggestion
func (r *KubeadmConfigReconciler) reconcileProxy(ctx context.Context, cfg *bootstrapv1.KubeadmConfig) ([]bootstrapv1.File, error) {
```
given that there are no proxy files?

File: bootstrap/kubeadm/api/v1alpha4/kubeadmbootstrapconfig_types.go (hunk 2 above)
Comment (fabriziopandini) [reply bootstrap/kubeadm/api/v1alpha4/kubeadmbootstrapconfig_types.go:350]: well, looking at the code below it seems this actually links to secrets, but it is not easy to understand this from the name/(missing) comments 😓

File: bootstrap/kubeadm/controllers/kubeadmconfig_controller.go [hunk 4]
```diff
@@ -563,6 +583,79 @@ func (r *KubeadmConfigReconciler) joinControlplane(ctx context.Context, scope *S
...
+			}
+			return nil, errors.Wrapf(err, "failed to retrieve Secret %q", key)
+		}
+	}
+	// resolve urls
+	httpProxyUrl, err := resolveProxyUrl(cfg.Spec.Proxy.HttpProxy, httpAuth)
+	if err != nil {
+		return nil, err
+	}
+	httpsProxyUrl, err := resolveProxyUrl(cfg.Spec.Proxy.HttpProxy, httpAuth)
```
Comment (fabriziopandini) [inline bootstrap/kubeadm/controllers/kubeadmconfig_controller.go:636]: I guess this should be
```suggestion
	httpsProxyUrl, err := resolveProxyUrl(cfg.Spec.Proxy.HttpsProxy, httpsAuth)
```

File: bootstrap/kubeadm/controllers/kubeadmconfig_controller.go [hunk 5]
```diff
@@ -563,6 +583,79 @@ func (r *KubeadmConfigReconciler) joinControlplane(ctx context.Context, scope *S
 	return ctrl.Result{}, nil
 }
 
+// resolveProxyUrl adds the user information to the proxy url
+func resolveProxyUrl(proxyUrl string, secret *corev1.Secret) (string, error) {
+	if secret == nil {
+		return proxyUrl, nil
+	}
+	username := secret.StringData["username"]
+	password := secret.StringData["password"]
```
Comment (fabriziopandini) [inline bootstrap/kubeadm/controllers/kubeadmconfig_controller.go:592]: What if the secret is malformed (without such fields). Should we have some validation and eventually return some meaningful error messages?

File: bootstrap/kubeadm/controllers/kubeadmconfig_controller.go [hunk 6]
```diff
@@ -563,6 +583,79 @@ func (r *KubeadmConfigReconciler) joinControlplane(ctx context.Context, scope *S
 	return ctrl.Result{}, nil
 }
 
+// resolveProxyUrl adds the user information to the proxy url
+func resolveProxyUrl(proxyUrl string, secret *corev1.Secret) (string, error) {
+	if secret == nil {
+		return proxyUrl, nil
```
Comment (fabriziopandini) [inline bootstrap/kubeadm/controllers/kubeadmconfig_controller.go:589]: Should we try to parse the url no matter if the secret is empty or not?

File: bootstrap/kubeadm/controllers/kubeadmconfig_controller.go [hunk 7]
```diff
@@ -563,6 +583,79 @@ func (r *KubeadmConfigReconciler) joinControlplane(ctx context.Context, scope *S
...
+	proxyDropin, err := proxy.Parse()
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed parseing proxy template")
+	}
+	// render files
+	proxyFile := bootstrapv1.File{
+		Path:    "/etc/systemd/system/containerd.service.d/http-proxy.conf",
+		Content: string(proxyDropin),
+	}
+	collected = append(collected, proxyFile)
```
Comment (fabriziopandini) [inline bootstrap/kubeadm/controllers/kubeadmconfig_controller.go:654]: Is there a case when we are collecting more than one file? if not, should we simplify this and return `(bootstrapv1.File, error)` without using a slice?

File: bootstrap/kubeadm/controllers/kubeadmconfig_controller.go [hunk 8]
```diff
@@ -563,6 +583,79 @@ func (r *KubeadmConfigReconciler) joinControlplane(ctx context.Context, scope *S
...
+	}
+	proxyDropin, err := proxy.Parse()
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed parseing proxy template")
+	}
+	// render files
+	proxyFile := bootstrapv1.File{
+		Path:    "/etc/systemd/system/containerd.service.d/http-proxy.conf",
+		Content: string(proxyDropin),
+	}
```
Comment (fabriziopandini) [inline bootstrap/kubeadm/controllers/kubeadmconfig_controller.go:653]: What about moving this in the dropin package and implement something like NewHtttpProxyDropin() that returns a file, so all the plugin details (e.g. the fact that the dropin internally uses a template, or the path where the dropin should be stored) are internal to this package?

File: bootstrap/kubeadm/api/v1alpha4/kubeadmbootstrapconfig_types.go [hunk 9]
```diff
@@ -333,5 +336,19 @@ type Filesystem struct {
 	ExtraOpts []string `json:"extraOpts,omitempty"`
 }
 
+// Proxy defines the http proxy settings used by container runtime during bootstrap
+type Proxy struct {
```
Comment (fabriziopandini) [inline bootstrap/kubeadm/api/v1alpha4/kubeadmbootstrapconfig_types.go:340]: Should we implement some webhook validation rules for this type?

Comment (fabriziopandini) [reply bootstrap/kubeadm/api/v1alpha4/kubeadmbootstrapconfig_types.go:340]: TBD if the same should be implemented for the copy of this type embedded in KCP

**Comment by hanlins** [comment]
> /milestone v0.4.0
> @hanlins are you still interested in working on this?

Sorry was on some tasks within my org recently, yes I will continue on this change shortly!

**Comment by vincepri** [comment]
/close

For lack of updates

**Pull request closed without merging**